File: find_minimum_in_sorted_rotated_array.cpp
//Time Complexity: O(LogN)
//Space Complexity: O(1)

#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.size()==0)
            return -1;
        int left=0;
        int right=nums.size()-1;
        while(left<right){
            int mid=left+(right-left)/2;
            if(nums[mid]>nums[right])
                left=mid+1;
            else
                right=mid;
        }
        return nums[left];
    }   
};

int main()
{
    Solution s;
    vector<int> nums{ 5, 7, 7, 8, 8, 10 };
    cout << s.findmin(nums); << endl;
    return 0;
}

File: FindMinimumInSortedRotatedArray.java
//TC : o(log n)
//SC: o(1)

class Solution {
    public int findMin(int[] nums) {
        int low = 0;
        int high = nums.length-1;
        if(nums[low] <= nums[high]){
            return nums[low];
        }
        while(low <= high){
            int mid = low + (high-low)/2;
            if(mid >0 && nums[mid-1] > nums[mid]){
                return nums[mid];
            }
            else if(nums[mid] > nums[high]){ // left array is sorted
                low = mid + 1;
            }
            else{
                high = mid -1;
            } 
        }
        return 89;
    }
}